File: optimizers/Adagrad.py
from .Optimizer import Optimizer, np


class Adagrad(Optimizer):
    def __init__(self, initial_x: np.ndarray, function, lr: float = 0.01, eps: float = 1e-08) -> None:
        params = dict(lr=lr, eps=eps)
        super().__init__(initial_x, function, params)

    def next_point(self) -> tuple[np.ndarray, float]:
        gradient = self.function.grad(self.x)
        self.acc += gradient**2
        adaptive_lr = self.params["lr"] / (np.sqrt(self.acc) + self.params["eps"])
        next_x = self.x - adaptive_lr * gradient
        return self.move_next(next_x)

    def reset(self) -> None:
        super().reset()
        self.acc: np.ndarray = np.zeros([2])

File: web/src/shared/lib/optimization-engine/optimizers/adagrad.ts
import type { OptimizerDescriptor } from "./types";
import { zero2, type Vec2 } from "@shared/lib/optimization-engine/linalg";
import { gradient } from "@shared/lib/optimization-engine/functions";

export const adagradOptimizer: OptimizerDescriptor = {
  name: "Adagrad",
  params: {
    lr: { default: 0.01, description: "базовая скорость обучения, делится на корень из суммы квадратов всех прошлых градиентов" },
    eps: { default: 1e-8, description: "малая добавка в знаменатель для численной устойчивости, настраивать обычно не нужно" },
  },
  createInstance(fn, initialX, params) {
    let x: Vec2 = initialX;
    let acc: Vec2 = zero2();
    return {
      get x() { return x; },
      params,
      next() {
        const grad = gradient(fn, x[0], x[1]);
        acc = [acc[0] + grad[0] ** 2, acc[1] + grad[1] ** 2];
        const adaptiveLr: Vec2 = [
          params.lr / (Math.sqrt(acc[0]) + params.eps),
          params.lr / (Math.sqrt(acc[1]) + params.eps),
        ];
        x = [x[0] - adaptiveLr[0] * grad[0], x[1] - adaptiveLr[1] * grad[1]];
        return { x, value: fn(x[0], x[1]) };
      },
      reset() {
        x = initialX;
        acc = zero2();
      },
    };
  },
};
